unittests/DXILLib/DxilLibTest.cpp

```diff
@@ -0,0 +1,152 @@
+
+  ASSERT_TRUE(
+      SUCCEEDED(compiler->Compile(sourceBlob, L"hlsl.hlsl", L"main", L"ps_6_0",
+                                  nullptr, 0, nullptr, 0, nullptr, &result)));
+  HRESULT status;
+  ASSERT_TRUE(SUCCEEDED(result->GetStatus(&status)));
+  ASSERT_TRUE(SUCCEEDED(status));
+  *outResult = result.Detach();
+}
+
+// Check that dxil lib can be loaded by default when creating
+// the compiler library. If the external DXIL Library isn't found, a warning is
+// issued. Test that there is no such warning.
+TEST(DxilLibTest, LoadFromCompiler) {
+
+  // Preloading validator lib so dxcompiler.dll will find it without RPATH's
+  // help Only Linux needs this, but we <3 Linux, so we help it out sometimes
+  dxc::DxcDllSupport validatorLib;
+  ASSERT_TRUE(SUCCEEDED(
+      validatorLib.InitializeForDll(dxc::kDxilLib, "DxcCreateInstance")))
+      << "Couldn't find " << dxc::kDxilLib;
+
+  dxc::DxcDllSupport compilerLib;
+  ASSERT_TRUE(SUCCEEDED(compilerLib.Initialize()));
+
+  // Compile trivial, valid shader and check for warnings
+  CComPtr<IDxcOperationResult> result;
+  CompileSomething(compilerLib, &result);
+
+  CComPtr<IDxcBlobEncoding> errBuf;
+  ASSERT_TRUE(SUCCEEDED(result->GetErrorBuffer(&errBuf)));
+
+  EXPECT_EQ(errBuf->GetBufferSize(), 0U)
+      << "Unexpected warning found in compiler error buffer: "
+      << (const char *)errBuf->GetBufferPointer();
+}
+
+// Check that dxil lib can be loaded and trivially used by a validator object
+TEST(DxilLibTest, LoadFromValidator) {
+  // Still need a compiler object to give us something to validate
+
+  dxc::DxcDllSupport validatorLib;
+  CComPtr<IDxcValidator> validator1;
+  CComPtr<IDxcValidator> validator2;
+  ASSERT_TRUE(SUCCEEDED(
+      validatorLib.InitializeForDll(dxc::kDxilLib, "DxcCreateInstance")))
+      << "Couldn't find " << dxc::kDxilLib;
+  EXPECT_TRUE(validatorLib.HasCreateWithMalloc());
+  ASSERT_TRUE(
+      SUCCEEDED(validatorLib.CreateInstance(CLSID_DxcValidator, &validator1)));
+
+  HeapMalloc allocator;
+  ASSERT_TRUE(SUCCEEDED(validatorLib.CreateInstance2(
+      &allocator, CLSID_DxcValidator, &validator2)));
+
+  // Create a simple compiled output to validate
+  dxc::DxcDllSupport compilerLib;
+  ASSERT_TRUE(SUCCEEDED(compilerLib.Initialize()));
+  CComPtr<IDxcOperationResult> compileResult;
+  CompileSomething(compilerLib, &compileResult);
+  CComPtr<IDxcBlob> blob;
+  ASSERT_TRUE(SUCCEEDED(compileResult->GetResult(&blob)));
+
+  CComPtr<IDxcOperationResult> validationResult1;
+  HRESULT status1;
+  CComPtr<IDxcBlobEncoding> errBuf1;
+  ASSERT_TRUE(SUCCEEDED(validator1->Validate(blob, DxcValidatorFlags_Default,
+                                             &validationResult1)));
+  ASSERT_TRUE(SUCCEEDED(validationResult1->GetStatus(&status1)));
+  EXPECT_TRUE(SUCCEEDED(status1));
+  ASSERT_TRUE(SUCCEEDED(validationResult1->GetErrorBuffer(&errBuf1)));
+  EXPECT_EQ(errBuf1->GetBufferSize(), 0U);
+
+  CComPtr<IDxcOperationResult> validationResult2;
```

Comment:
What is the difference between validationResult1 and validationResult2?

Reply:
One is created by the plain `validator1` that uses the `CreateInstance` API call and the other is created by `validator2` which uses the `CreateInstance2` API call that also provides a custom allocator. I kept the variables used fully separate to minimize potential for interference between them like a potential stale value from an earlier call.